phpBB/phpbb/style/manager.php

```diff
@@ -0,0 +1,436 @@
+*
+* @copyright (c) phpBB Limited <https://www.phpbb.com>
+* @license GNU General Public License, version 2 (GPL-2.0)
+*
+* For full copyright and license information, please see
+* the docs/CREDITS.txt file.
+*
+*/
+
+namespace phpbb\style;
+
+use phpbb\style\exception;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+* The extension manager provides means to activate/deactivate extensions.
+*/
+class manager
+{
+	protected $cache;
+
+	/** @var ContainerInterface */
+	protected $container;
+
+	protected $db;
+	protected $config;
+	protected $filesystem;
+	protected $styles_table;
+	protected $users_table;
+	protected $phpbb_root_path;
+
+	/** @var \phpbb\textformatter\cache_interface */
+	protected $text_formatter_cache;
+
+	protected $reserved_style_names = array('adm', 'admin', 'all');
+
+	protected $styles_path;
+	protected $styles_path_absolute = 'styles';
+
+	public function __construct(\phpbb\cache\service $cache = null, \phpbb\config\config $config, ContainerInterface $container, \phpbb\db\driver\driver_interface $db, \phpbb\filesystem\filesystem_interface $filesystem, $styles_table, $users_table, $phpbb_root_path)
+	{
+		$this->cache = $cache;
+		$this->config = $config;
+		$this->container = $container;
+		$this->db = $db;
+		$this->filesystem = $filesystem;
+		$this->styles_table = $styles_table;
+		$this->users_table = $users_table;
+		$this->phpbb_root_path = $phpbb_root_path;
+
+		$this->text_formatter_cache = $container->get('text_formatter.cache');
+		$this->default_style = $config['default_style'];
+		$this->styles_path = $this->phpbb_root_path . $this->styles_path_absolute . '/';
+	}
+
+
+	public function install($dir)
+	{
+		if (in_array($dir, $this->reserved_style_names))
+		{
+			throw new exception('STYLE_NAME_RESERVED');
+		}
+
+		if ($this->get_style_data('style_path', $dir))
+		{
+			throw new exception('STYLE_ALREADY_INSTALLED');
+		}
+
+		$cfg = $this->read_style_cfg($dir);
+
+		if (!$cfg)
+		{
+			throw new exception('STYLE_FOLDER_INVALID');
+		}
+
+		if ($this->get_style_data('style_name', $cfg['name']))
+		{
+			throw new exception('STYLE_ALREADY_INSTALLED');
+		}
+
+		// Style should be available for installation
+		$sql_ary = array(
+			'style_name'		=> $cfg['name'],
+			'style_copyright'	=> $cfg['copyright'],
+			'style_active'		=> 1,
+			'style_path'		=> $dir,
+			'bbcode_bitfield'	=> $cfg['template_bitfield'],
+			'style_parent_id'	=> 0,
+			'style_parent_tree'	=> '',
+		);
+
+		if ($cfg['parent'])
+		{
+			$parent_data = $this->get_style_name('style_name', $cfg['parent']);
+			if ($parent_data)
+			{
+				$sql_ary['style_parent_id'] = $parent_data['style_id'];
+				$sql_ary['style_parent_tree'] = $parent_data['tree'];
+			}
+		}
+
+		$sql = 'INSERT INTO ' . $this->styles_table . '
+			' . $this->db->sql_build_array('INSERT', $sql_ary);
+
+		if (!$this->db->sql_query($sql))
+		{
+			throw new exception('STYLE_NOT_INSTALLED');
+		}
+
+		$this->text_formatter_cache->invalidate();
```

Comment:
Not for later, this should be moved to a listener (do not do it now, we will do that later in another PR)